File: Delete Middle Node of Linked List.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow_prev = None
        slow = head
        fast = head

        while fast is not None and fast.next is not None:
            fast = fast.next.next
            slow_prev = slow
            slow = slow.next
        if slow_prev is not None:
            slow_prev.next = slow.next
            return head
        if head is None:
            return head 
        return head.next
        

File: cpp/Delete Middle Node of Linked List.cpp
#include "common.h"


struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* deleteMiddle(ListNode* head) {
        ListNode * slow_prev = nullptr;
        ListNode * slow = head;
        ListNode * fast = head;

        while (fast && fast->next) {
            fast = fast->next->next;
            slow_prev = slow;
            slow = slow->next;
        }
        if (slow_prev) {
            slow_prev->next = slow->next;
            return head;
        }
        if (!head) {
            return head;
        }
        return head->next;
    }
};